src/Microsoft.AspNet.OData.Shared/Query/Expressions/SelectExpandBinder.cs

```diff
@@ -152,40 +153,116 @@ internal Expression CreatePropertyNameExpression(IEdmEntityType elementType, IEd
             return Expression.Constant(property.Name);
         }
 
-        internal Expression CreatePropertyValueExpression(IEdmEntityType elementType, IEdmProperty property, Expression source)
+        internal Expression CreatePropertyValueExpression(IEdmStructuredType elementType, IEdmProperty property, Expression source)
         {
-            Contract.Assert(elementType != null);
             Contract.Assert(property != null);
             Contract.Assert(source != null);
 
-            return CreatePropertyValueExpressionWithFilter(elementType, property, source, filterClause: null);
+            return CreatePropertyValueExpressionWithFilter(elementType, property, source, null);
         }
 
-        internal Expression CreatePropertyValueExpressionWithFilter(IEdmEntityType elementType, IEdmProperty property,
-            Expression source, FilterClause filterClause)
+        internal Expression CreatePropertyValueExpressionWithFilter(IEdmStructuredType elementType, IEdmProperty property,
+            Expression source, ExpandedReferenceSelectItem expandItem)
         {
             Contract.Assert(elementType != null);
             Contract.Assert(property != null);
             Contract.Assert(source != null);
 
-            IEdmEntityType declaringType = property.DeclaringType as IEdmEntityType;
-            Contract.Assert(declaringType != null, "only entity types are projected.");
+            FilterClause filterClause = expandItem != null ? expandItem.FilterOption : null;
+
+            IEdmStructuredType declaringType = property.DeclaringType;
+
+            //if there is a path in $expand then find the right declaring type;
+            if (expandItem != null)
+            {
+                foreach (ODataPathSegment segment in expandItem.PathToNavigationProperty)
+                {
+                    PropertySegment propertyAccessPathSegment = segment as PropertySegment;
+                    if (propertyAccessPathSegment != null)
+                    {
+                        IEdmProperty newProperty = propertyAccessPathSegment.Property;
+                        declaringType = newProperty.DeclaringType;
+                        break;
+                    }
+                    else
+                    {
+                        TypeSegment castSegment = segment as TypeSegment;
+                        if (castSegment != null)
+                        {
+                            continue;
+                        }
+
+                        break;
```

Comment:
>break [](start = 24, length = 5)

so, if we find a segment other than a PropertySegment or a TypeSegment, we just ignore the entire path?  This would be an error, right?